src/Microsoft.OData.Edm/ExtensionMethods/AnnotationsTargetHelper.cs

```diff
@@ -0,0 +1,205 @@
+                }
+                else
+                {
+                    return singleton;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmContainerElement(IEdmModel model, List<IEdmElement> results, IEdmEntityContainerElement edmEntityContainerElement, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            // .../MyEntitySet/MySchema.MyEntityType/...
+            // .../MyEntitySet/MyComplexProperty/...
+            // .../MySingleton/MyComplexProperty/...
+            IEdmSchemaType type = model.FindType(targetSegments[level]);
+
+            if (type != null)
+            {
+                results.Add(type);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmSchemaType(model, results, type, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    // If we have this path MySchema.MyEntityContainer/MyEntitySet/MySchema.MyEntityType
+                    // Should we return null since it's invalid?
+                    return type;
+                }
+            }
+
+            IEdmType edmType = null;
+            IEdmStructuredType structuredType;
+
+            if (edmEntityContainerElement is IEdmEntitySet edmEntitySet)
+            {
+                edmType = edmEntitySet.Type.AsElementType();
+            }
+
+            if (edmEntityContainerElement is IEdmSingleton edmSingleton)
+            {
+                edmType = edmSingleton.Type.AsElementType();
+            }
+
+            if ((structuredType = edmType as IEdmStructuredType) != null)
+            {
+                // .../MyEntitySet/MyProperty/...
+                // .../MyEntitySet/MyNavigationProperty/...
+                // .../MySingleton/MyProperty/...
+                // .../MySingleton/MyNavigationProperty/...
+                IEdmProperty property = structuredType.FindProperty(targetSegments[level]);
+
+                if (property != null)
+                {
+                    results.Add(property);
+
+                    if (targetSegmentsCount - 1 > level)
+                    {
+                        return HandleIEdmProperty(model, results, property, targetSegments, targetSegmentsCount, level + 1);
+                    }
+
+                    return property;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmSchemaType(IEdmModel model, List<IEdmElement> results, IEdmSchemaType edmSchemaType, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            IEdmStructuredType structuredType;
+
+            if ((structuredType = edmSchemaType as IEdmStructuredType) != null)

```

Comment:
>if ((structuredType = edmSchemaType as IEdmStructuredType) != null) [](http://example.com/codeflow?start=12&length=67)

a lot of duplicated codes in this helper class. For example:
Line 141 is checked at Line 113.